File: swp_loan/swp_loan/page/swp_file_attachment/swp_file_attachment.js
// import CustomFileUploaderWrapper from "./custom_file_uploader.js";

frappe.pages["swp-file-attachment"].on_page_load = function (wrapper) {
	const can_create = frappe.model.can_create("Attachment");
	const can_delete = frappe.model.can_delete("Attachment");
	const can_select = frappe.model.can_select("Attachment");
	const can_export = frappe.model.can_export("Attachment");
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'swp-file-attachment',
		single_column: true
	});

    // Get parent name
    const parent_name = frappe.get_route_str().split("/")[1];
	async function loadLoanAndAttachment() {
    let loan = {
        customer_name: "No data",
        contract_number: "No data",
        collateral_number: "No data"
    };

    // Load loan application
    const docs = await frappe.db.get_list("Loan Application", { fields: ["name"], limit: 1 });
    if (docs.length > 0) {
        const full_doc = await frappe.db.get_doc("Loan Application", docs[0].name);
		console.log(full_doc);
        loan = {
            customer_name: full_doc.cus_first_name + " " + full_doc.cus_last_name,
            contract_number: full_doc.cus_addresses_detail?.[0]?.mobile_number || "No mobile number",
            collateral_number: full_doc.cus_identification_number,
        };
    }
    $(frappe.render_template("swp_file_attachment", { loan })).appendTo(page.main);

    // Load attachment
    const doc = await frappe.db.get_list("Attachment", {
        fields: ["*"],
        filters: {
            is_group: 1,
            parent_attachment: "",
            is_custom: 0,
            reference_document: parent_name
        }
    });

    if (doc.length > 0) {
        set_tree(doc[0].name, doc[0].description);
    } else {
        const result_root = await frappe.call({
            method: "z_loan.loan.page.file_attachment.file_attachment.create_root",
            args: { docname: parent_name }
        });
        create_folder(parent_name, "Home/Attachments", async () => {
            const r = await frappe.call({
                method: "z_loan.api.generate_folder_structure",
                args: {
                    reference_document: parent_name,
                    loan_app_doc: frappe.route_options.loan_app_doc,
                    sales_kit_doc: frappe.route_options.sales_kit_doc,
                }
            });
            set_tree(result_root.message.name, result_root.message.description);
        });
    }
}

loadLoanAndAttachment();


	frappe.ui.keys.add_shortcut({
		shortcut: "up",
        action: () => handle_navigation_node("up"),
        description: __("Navigate node up"),
        page: this.page
	});
	
	frappe.ui.keys.add_shortcut({
		shortcut: "down",
        action: () => handle_navigation_node("down"),
        description: __("Navigate node down"),
        page: this.page
	});

	if(can_export){
		page.add_inner_button(__("Downloads"), function () {
			download_files(parent_name);
		});
	}

	function set_tree(root_name,root_description) {
		doctype = "Attachment"
		parent = ""
		frappe.file_attachment_tree = new frappe.ui.Tree({
			parent: $("#treeview"),
			label:root_description,
			root_value: root_name,
			method: "z_loan.loan.page.file_attachment.file_attachment.get_children",
			args: {
				doctype: doctype
			},
			get_label: function (node) {
				let node_label = "";
				if(node.title) {
					node_label = node.title
				}
				else {
					node_label = root_description
				}
				if (node.data.is_required) {
					node_label += ' <b style="color:red;">*</b>'
				}
				if (node.data.is_verified) {
					node_label += " ✅"
				}
				return node_label
			},
			toolbar: [
				{
					label: __("New Folder"),
					condition: function (node) {
						return node.expandable == 1 && can_create
					},
					click: function (node) {
						add_node(node)
					}
				},
				{
					label: __("Add File"),
					condition: function (node) {
						return node.expandable == 1 && (node.is_root === "undefined" || !node.is_root) && can_create
					},
					click: function (node) {
						add_file(node)
					}
				},
				{
					label: __("Verify"),
					condition: function (node) {
						return node.expandable == 0 && can_select && !node.data.is_verified
					},
					click: function (node) {
						verify_node(node)
					}
				},
				{
					label: __("Move",null,"Attachment"),
					condition: function (node) {
						return node.expandable == 0 && can_create
					},
					click: function (node) {
						move_node(node)
					}
				},
				{
					label: __("Delete"),
					condition: function (node) {
						return !node.is_root && node.data.is_custom && can_delete
					},
					click: function (node) {
						delete_node(node)

					}
				}
			],
			on_click: (node) => {
				parent = node["name"]
				if (node["data"]["expandable"] == 0) {
					set_preview(doctype, node["data"]["value"])
				}
				else if (node["data"]["expandable"] == undefined) {
					set_preview_thumbnail()
				}
			}
		});
		frappe.file_attachment_tree.refresh();
	}

	function set_preview(doctype, parent) {
		frappe.call({
			method: "frappe.client.get",
			args: {
				doctype: doctype,
				name: parent
			},
			callback: function (response) {
				var doc = response.message;
				if (doc) {
					frappe.call({
						method: "z_loan.loan.page.file_attachment.file_attachment.get_file_by_url",
						args: { file_url: doc.attach_file },
						callback: function (res) {
							doc_file = res.message[0]
							let preview = $("#preview");
							let $preview = ""
							let file_extension = doc_file.file_type.toLowerCase();
							if (frappe.utils.is_image_file(doc_file.file_url)) {
								$preview = $(`<div class="img_preview">
												<img
													class="img-responsive"
													src="${frappe.utils.escape_html(doc_file.file_url)}"
												/>
											</div>`);
							} else if (frappe.utils.is_video_file(doc_file.file_url)) {
								$preview = $(`<div class="img_preview">
												<video width="480" height="320" controls>
													<source src="${frappe.utils.escape_html(doc_file.file_url)}">
													${__("Your browser does not support the video element.")}
												</video>
											</div>`);
							} else if (file_extension === "pdf") {
								$preview = $(`<div class="img_preview">
												<object style="background:#323639;" width="100%">
													<embed
														style="background:#323639;"
														width="100%"
														height="690"
														src="${frappe.utils.escape_html(doc_file.file_url)}" type="application/pdf"
													>
												</object>
											</div>`);
							} else if (file_extension === "mp3") {
								$preview = $(`<div class="img_preview">
												<audio width="480" height="60" controls>
													<source src="${frappe.utils.escape_html(doc_file.file_url)}" type="audio/mpeg">
													${__("Your browser does not support the audio element.")}
												</audio >
											</div>`);
							}

							if ($preview) {
								preview.html($preview);
							}
						}
					})
				}
			}
		});
	}
	async function set_preview_thumbnail() {
		let preview = $('#preview');
		if(!frappe.all_node){
			await get_all_node();
		}
		preview.empty()
		frappe.all_node.forEach(function (folder) {
			let $folderDiv = $('<div class="folder"></div>');
			let $folderHeading = $("<h3></h3>").text(folder.folder);
			$folderDiv.append($folderHeading);

			folder.url.forEach(function (url) {
				let $preview = $(`<div class="img_thumbnail">
											<img
												class="img-responsive"
												src="${frappe.utils.escape_html(url)}"
											/>
										</div>`);
				$folderDiv.append($preview);
			});
			preview.append($folderDiv);
		});
	}

	function add_file(node) {
		root_path = get_root_path(node);
		new frappe.ui.FileUploader({
			folder: "Home/Attachments/" + parent_name + "/" + root_path,
			make_attachments_public: 0,
			on_success: (file) => {
				frappe.call({
					method: "z_loan.loan.page.file_attachment.file_attachment.create_attachment",
					args: {
						docname: parent_name,
						description: file.file_name,
						is_custom: 1,
						is_group: 0,
						parent_attachment: node.data.value,
						file: file
					},
					callback: function (r) {
						refresh_tree();
						frappe.all_node = null;
					}
				});
			},
		});
	}
	function add_node(node) {
		let title = node.title ? node.title : node.label;
		var dialog = new frappe.ui.Dialog({
			title: __("{0}", [title]),
			fields: [
				{
					label: __("Folder Name"),
					fieldname: "folder_name",
					fieldtype: "Data"
				},
			]
		});
		dialog.set_primary_action(__("New Folder"), function () {
			var d = dialog.get_values();
			frappe.call({
				method: "z_loan.loan.page.file_attachment.file_attachment.create_attachment",
				args: {
					docname: parent_name,
					description: d.folder_name,
					is_custom: 1,
					is_group: 1,
					parent_attachment: node.data.value,
					file: null
				},
				callback: function (r) {
					root_path = get_root_path(node);
					file_path = "Home/Attachments/" + parent_name + (root_path === "" ? "" : "/" + root_path);
					create_folder(d.folder_name, file_path, () => {
						dialog.hide();
						refresh_tree();
					});
				}
			});

		});
		dialog.show();
	}
	function delete_node(node) {
		var dialog = new frappe.ui.Dialog({
			title: __("{0}", [node.title]),
			fields: [
				{
					label: __("Are you sure you want to delete this document?"),
					fieldname: "confirm_message",
					fieldtype: "HTML",
					options: `<p>${__("Are you sure you want to delete this document? This action cannot be undone.")}</p>`
				}
			]
		});
		dialog.set_primary_action(__("Delete"), function () {
			var d = dialog.get_values();
			frappe.call({
				method: "z_loan.loan.page.file_attachment.file_attachment.delete_doc",
				args: {
					doctype: doctype,
					docname: node.data.value
				},
				callback: function (r) {
					node.parent.remove();
					dialog.hide();
					frappe.all_node = null;
					set_preview_thumbnail();
				}
			});

		});
		dialog.show();
	}
	function verify_node(node) {
		var dialog = new frappe.ui.Dialog({
			title: __("{0}", [node.title]),
			fields: [
				{
					label: __("Document Verification"),
					fieldname: "verify_message",
					fieldtype: "HTML",
					options: `<p>${__("Please verified the document and confirmed that all the information is accurate?")}</p>`
				}
			]
		});
		dialog.set_primary_action(__("Verify"), function () {
			frappe.call({
				method: "z_loan.loan.page.file_attachment.file_attachment.verify_doc",
				args: {
					doctype: doctype,
					docname: node.data.value
				},
				callback: function (r) {
					dialog.hide();
					frappe.file_attachment_tree.refresh()
				}
			});

		});
		dialog.show();
	}
	function move_node(node) {
		folder_list = [];
		Object.keys(frappe.file_attachment_tree.nodes).forEach(key => {
			node_item = frappe.file_attachment_tree.nodes[key];
			if (node_item.expandable == true && (node_item.is_root == undefined || node_item.is_root == false) &&
				node.parent_node.title != node_item.title) {
				folder_list.push({
					"value":node_item.data.value,
					"label":node_item.title
				});
			}
		});
		var dialog = new frappe.ui.Dialog({
			title: __("{0}", [node.title]),
			fields: [
				{
					label: __("Move to folder"),
					fieldname: "new_folder",
					fieldtype: "Select",
					options: folder_list
				},
			]
		});
		dialog.set_primary_action(__("Move"), function () {
			let d = dialog.get_values();
			frappe.call({
				method: "z_loan.loan.page.file_attachment.file_attachment.move_file",
				args: {
					docname: node.data.value,
					new_parent: d.new_folder
				},
				callback: function (r) {
					dialog.hide();
					refresh_tree();
				}
			});
		});
		dialog.show();
	}
	function create_folder(folder_name, root_folder, call_back) {
		if (folder_name.indexOf("/") > -1) {
			frappe.throw(__("Folder name should not include "/" (slash)"));
		}
		const data = {
			file_name: folder_name,
			folder: root_folder,
		};
		frappe.call({
			method: "frappe.core.api.file.create_new_folder",
			args: data,
			callback: call_back
		});
	}

	function get_root_path(node) {
		const parents = [];
		if (node.parent_node != undefined) {
			parents.push(node.title);
			let current_node = node.parent_node;
			while (current_node.parent_node != undefined) {
				parents.push(current_node.title);
				current_node = current_node.parent_node;
			}
		}
		return parents.reverse().join("/");
	}
	function get_all_folder_by_root(node) {
		const parents = [];
		parents.push(node.title);
		let current_node = node.parent_node;
		while (current_node.parent_node != undefined) {
			parents.push(current_node.title);
			current_node = current_node.parent_node;
		}
		return parents.reverse().join("/");
	}
	function refresh_tree() {
		frappe.file_attachment_tree.refresh();
	}
	function get_all_node(){
		return frappe.call({
			method: 'z_loan.api.get_all_node',
			args: {
				reference_document: parent_name
			},
			callback: function (response) {
				frappe.all_node = response.message.node_list;
			}
		});
	}
	function handle_navigation_node(direction){
		const current_node = $(document.activeElement).find(".tree-link.selected");
		const all_node = $(document.activeElement).find(".tree-link");
		const all_node_not_folder_opened = all_node.filter(function() {
			return $(this).parent().find(".tree-node.opened").length === 0;
		});
		const current_index = all_node_not_folder_opened.index(current_node);
		let next_index = 0
		if(direction === "down"){
			next_index = current_index + 1;
		}
		else{
			next_index = current_index - 1;
		}
		if (next_index >= 0 && next_index < all_node_not_folder_opened.length) {
			next_node = all_node_not_folder_opened.eq(next_index);
			next_node.click();
		}
	}
	function download_files(reference_document){
		open_url_post("/api/method/z_loan.api.download_zip_files", {
            reference_document: reference_document
          });
	}
};

File: swp_loan/swp_loan/page/swp_file_attachment/swp_file_attachment.py


import frappe
from frappe import _
import json
from frappe.core.api.file import create_new_folder
from frappe.query_builder import DocType,functions as fn

@frappe.whitelist()
def get_file_by_url(file_url):
    return frappe.get_list(doctype='File', fields=['*'], filters={'file_url': file_url}, ignore_permissions=True)


@frappe.whitelist()
def delete_doc(doctype, docname):
    if not frappe.db.exists(doctype, docname):
        frappe.throw(_('Document {0} {1} does not exist').format(doctype, docname))
    try:
        attachment = frappe.get_doc(doctype, docname)
        linked_folder_list = frappe.get_list(doctype, fields=['name'],filters=[['old_parent','=',docname]],ignore_permissions=True)
        frappe.db.set_value(doctype, [item['name'] for item in linked_folder_list], 'old_parent', None)
        attachment.delete(ignore_permissions=True)
        if attachment.is_group:
            folder_doc = frappe.get_doc('File', {'is_folder': 1,'file_name':attachment.description})
            folder_doc.delete(ignore_permissions=True)
    except Exception as error:
        frappe.throw(_('Cannot delete this folder. The folder is not empty.'))


    return {'status': 'success', 'message': _('Document deleted successfully')}

@frappe.whitelist()
def verify_doc(doctype, docname):
    doc = frappe.get_doc(doctype, docname)
    doc.set('is_verified', 1)
    doc.save(ignore_permissions=True)
    return {'status': 'success', 'message': _('Verified document successfully')}

@frappe.whitelist()
def move_file(docname,new_parent):
    doc = frappe.get_doc('Attachment', docname)
    doc.set('old_parent', doc.parent_attachment)
    doc.set('parent_attachment', new_parent)
    doc.save(ignore_permissions=True)
    return {'status': 'success', 'message': _('Moved document successfully')}

@frappe.whitelist()
def create_root(docname):
    root_attachment = frappe.new_doc('Attachment')
    root_attachment.reference_document = docname
    root_attachment.description = docname
    root_attachment.is_custom = 0
    root_attachment.is_group = 1
    root_attachment.save(ignore_permissions=True)
    return root_attachment

@frappe.whitelist()
def create_attachment(docname,description,is_custom,is_group,parent_attachment,file,is_required = 0):
    new_attachment = frappe.new_doc('Attachment')
    new_attachment.reference_document = docname
    new_attachment.description = description
    new_attachment.is_required = is_required
    new_attachment.is_custom = is_custom
    new_attachment.is_group = is_group
    new_attachment.parent_attachment = parent_attachment
    new_attachment.save(ignore_permissions=True)
    if file:
        if isinstance(file,str):
            # case create file from ui
            file = json.loads(file)
            update_file = frappe.get_doc('File', {'file_url': file['file_url']})
        elif isinstance(file,dict):
            # case copy file attachment
            file = frappe.get_doc('File', {'file_url': file['file_url']})
            update_file = frappe.copy_doc(file)
            from frappe.model.naming import make_autoname
            update_file.name = make_autoname('hash')
            update_file.flags.disable_watermark = True
            path_folder = update_file.folder.split('/')
            path_folder[2] = docname # path folder main parent attachment
            update_file.folder = '/'.join(path_folder)
        else:
            # case create file from generate folder structure
            update_file = file
        update_file.attached_to_doctype = 'Attachment'
        update_file.attached_to_field = 'attach_file'
        update_file.attached_to_name = new_attachment.name
        update_file.save(ignore_permissions=True)
        new_attachment.attach_file = update_file.file_url
        new_attachment.save(ignore_permissions=True)
    return new_attachment

def intial_location_structure(**kwargs):
    #region parameters
    parameters = {
        'reference_document' : None,
        'document_group' : None,
    }
    parameters.update(kwargs)
    #endregion parameters

    #region coding
    reference_document = parameters.get('reference_document')
    locations = []
    new_attachments = []
    main_parent_attachment = frappe.db.get_value('Attachment',{'reference_document':reference_document,'parent_attachment':['is',None]},'name')
    document_detail = frappe.db.sql('''
        select 
            document.location,
            document_group_detail.is_required
        from `tabDocument Group Detail` as document_group_detail
        inner join `tabDocument` as document on document.name = document_group_detail.document
        where parenttype = 'Document Group' and parentfield = 'document_group_detail' and document_group_detail.parent = %s
    '''
    ,(parameters.get('document_group'))
    ,as_dict=True)
    for document in document_detail:
        folder_name_list = document.location.split('/')
        for index,folder_name in enumerate(folder_name_list[1:]):
            locations.append({
                'location': document.location,
                'reference_document' : reference_document,
                'description' : folder_name,
                'is_required' : document.is_required if len(folder_name_list) == index + 2 else 0,
                'is_custom' : 0,
                'is_group' : 1,
                'parent_attachment' : folder_name_list[index] if index > 0 else None
            })
    locations = _delete_duplicates_attachment(locations,['location','is_required'])
    for location in locations:
        if location.get('parent_attachment'):
            for attachment in new_attachments:
                if attachment.description == location.get('parent_attachment'):
                    parent_attachment = attachment.name
                    break
        else:
            parent_attachment = main_parent_attachment
        new_attachment = create_attachment(
            docname = location.get('reference_document'),
            description = location.get('description'),
            is_required = location.get('is_required'),
            is_custom = location.get('is_custom'),
            is_group = location.get('is_group'),
            parent_attachment = parent_attachment,
            file = None)
        new_attachments.append(new_attachment)
        location_list = location.get('location').split('/')[1:]
        root_path = ''
        if location.get('parent_attachment'):
            for l in location_list:
                root_path += '/' + l
                if l == location.get('parent_attachment'):
                    break
        folder = 'Home/Attachments/'+location.get('reference_document') + root_path
        create_new_folder(file_name = location.get('description'), folder = folder)
    #endregion coding

    #region result
    result = {
        'locations' : new_attachments,
    }
    return result
    #endregion result

def _delete_duplicates_attachment(records, field_to_ignore):
    seen_keys = set()
    unique_records = []
    records = list({tuple(d.items()): d for d in records}.values())
    for record in records:
        key = tuple((key, record[key]) for key in record if key not in field_to_ignore)
        if key not in seen_keys:
            seen_keys.add(key)
            unique_records.append(record)
        else:
            for unique_record in unique_records:
                if record.get('description') == unique_record.get('description'):
                    unique_record['is_required'] = unique_record.get('is_required') if unique_record.get('is_required') == 1 else record.get('is_required')
    return unique_records

@frappe.whitelist()
def get_children(doctype, parent='', **filters):
	return _get_children(doctype, parent)


def _get_children(doctype, parent='', ignore_permissions=False):
	parent_field = 'parent_' + frappe.scrub(doctype)
	filters = [[f"ifnull(`{parent_field}`,'')", '=', parent], ['docstatus', '<', 2]]

	return frappe.get_list(
		doctype,
		fields=[
			'name as value',
			'description as title',
			'is_group as expandable',
            'is_required',
            'is_verified',
            'is_custom'
		],
		filters=filters,
		order_by='is_group desc,description',
		ignore_permissions=ignore_permissions,
	)

def get_all_node(**kwargs):
    #region parameters
    parameters = {
        'reference_document' : None,
    }
    parameters.update(kwargs)
    #endregion parameters

    #region coding
    reference_document = parameters['reference_document']
    node_list = []
    attachment_folder = DocType('Attachment')
    attachment_file = DocType('Attachment')
    query_attachment_list = (
        frappe.qb.from_(attachment_folder)
        .join(attachment_file).on((attachment_folder.name == attachment_file.parent_attachment) & (attachment_file.is_group == 0))
        .select(
            attachment_folder.description.as_('folder'),
            fn.GroupConcat(attachment_file.attach_file).as_('url'),
        )
        .where((attachment_folder.is_group == 1) & (attachment_folder.reference_document == reference_document))
        .groupby(attachment_folder.description).orderby(attachment_folder.description))
    result_attachment_list = query_attachment_list.run(as_dict=True)
    from frappe.utils import is_image
    for result_attachment in result_attachment_list:
        url_list = result_attachment['url'].split(',')
        url_img_list = [url for url in url_list if is_image(url)]
        if url_img_list:
            node_list.append({
                'folder': result_attachment['folder'],
                'url': url_img_list
            })
    #endregion coding

    #region result
    result = {
        'node_list' : node_list,
    }
    return result
    #endregion result

def copy_attachments_to_target(**kwargs):
    #region parameters
    parameters = {
        'attachment_list' : None,
        'from_reference_document' : None,
        'to_reference_document' : None,
    }
    parameters.update(kwargs)
    #endregion parameters

    #region coding
    attachment_list = parameters['attachment_list']
    from_reference_document = parameters['from_reference_document']
    to_reference_document = parameters['to_reference_document']
    old_attachment_list = frappe.db.get_list(doctype = 'Attachment',filters = [['reference_document','=',from_reference_document],['parent_attachment','is','set']],fields = ['name','description','is_required','is_verified','is_custom','is_group','parent_attachment','attach_file'],order_by='is_group desc,modified asc',limit_page_length=0)
    old_parent_attachment = None
    new_parent_attachment = None
    main_parent_attachment = frappe.db.get_value('Attachment',{'reference_document':to_reference_document,'parent_attachment':['is',None]},'name')
    file_attachment_list, folder_attachment_is_parent_list, folder_attachment_is_not_parent_list = mapping_attachment(old_attachment_list,attachment_list,main_parent_attachment)
    new_folder_attachment_list = []
    for folder_attachment in folder_attachment_is_parent_list:
        new_folder_attachment = copy_folder_attachment(folder_attachment,to_reference_document,main_parent_attachment,attachment_list)
        new_folder_attachment_list.append(new_folder_attachment)
        attachment_list.append(new_folder_attachment)
    for folder_attachment in folder_attachment_is_not_parent_list:
        new_folder_attachment = copy_folder_attachment(folder_attachment,to_reference_document,main_parent_attachment,attachment_list,new_folder_attachment_list)
        new_folder_attachment_list.append(new_folder_attachment)
        attachment_list.append(new_folder_attachment)
    for file_attachment in file_attachment_list:
        new_attachment = copy_file_attachment(file_attachment,old_parent_attachment,new_parent_attachment,old_attachment_list,attachment_list,to_reference_document)
        attachment_list.append(new_attachment)
    #endregion coding

    #region result
    result = {
        'attachment_list' : attachment_list,
    }
    return result
    #endregion result

def mapping_attachment(old_attachment_list,new_attachment_list,main_parent_attachment):
    file_attachment_list = []
    folder_attachment_is_parent_list = []
    folder_attachment_is_not_parent_list = []
    for old_attachment in old_attachment_list:
        if old_attachment.is_group == 1:
            is_match_folder = False
            for new_attachment in new_attachment_list:
                if new_attachment.description == old_attachment.description:
                    is_match_folder = True
                    break
            if not is_match_folder:
                is_match_parent = False
                parent_attachment_description = None
                for parent_attachment in old_attachment_list:
                    if parent_attachment.name == old_attachment.parent_attachment:
                        parent_attachment_description = parent_attachment.description
                        for new_parent_attachment in new_attachment_list:
                            if new_parent_attachment.description == parent_attachment.description:
                                is_match_parent = True
                                old_attachment.parent_attachment = new_parent_attachment.name
                                break
                        break
                if not is_match_parent:
                    if parent_attachment_description:
                        old_attachment.parent_attachment = parent_attachment_description
                        folder_attachment_is_not_parent_list.append(old_attachment)
                    else:
                        old_attachment.parent_attachment = main_parent_attachment
                        folder_attachment_is_parent_list.append(old_attachment)
                else:
                    folder_attachment_is_parent_list.append(old_attachment)
        elif old_attachment.is_group == 0:
            file_attachment_list.append(old_attachment)
    return file_attachment_list, folder_attachment_is_parent_list, folder_attachment_is_not_parent_list

def copy_folder_attachment(attachment,to_reference_document,main_parent_attachment,attachment_list,new_attachment_list = []):
    new_attachment = None
    for parent_attachment in new_attachment_list:
        if parent_attachment.description == attachment.parent_attachment:
            attachment.parent_attachment = parent_attachment.name
            break
    new_attachment = create_attachment(
        docname = to_reference_document,
        description = attachment.description,
        is_required = 0,
        is_custom = attachment.is_custom,
        is_group = attachment.is_group,
        parent_attachment = attachment.parent_attachment,
        file = None)
    root_path = ''
    root_path = search_location(new_attachment,attachment_list,main_parent_attachment,root_path,is_last_folder=True)
    folder = 'Home/Attachments/'+ new_attachment.reference_document + root_path
    create_new_folder(file_name = new_attachment.description, folder = folder)
    return new_attachment

def search_location(attachment,attachment_list,main_parent_attachment,root_path = '',is_last_folder = False):
    if not is_last_folder:
        root_path = '/' + attachment.description + root_path
    if attachment.parent_attachment == main_parent_attachment:
        return root_path
    else:
        for parent_attachment in attachment_list:
            if parent_attachment.name == attachment.parent_attachment:
                return search_location(parent_attachment,attachment_list,main_parent_attachment,root_path)

def copy_file_attachment(file_attachment,old_parent_attachment,new_parent_attachment,old_attachment_list,attachment_list,to_reference_document):
    if old_parent_attachment != file_attachment.parent_attachment:
        old_parent_attachment = file_attachment.parent_attachment
        is_parent_attachment = False
        for old_attachment in old_attachment_list:
            if old_attachment.name == file_attachment.parent_attachment:
                old_attachment_doc = old_attachment
                break
        for parent_attachment in attachment_list:
            if parent_attachment.description == old_attachment_doc.get('description'):
                new_parent_attachment = parent_attachment.name
                is_parent_attachment = True
                break
        if not is_parent_attachment:
            new_parent_attachment = attachment_list[0].name 
    file = {'file_url':file_attachment.attach_file}
    new_attachment = create_attachment(to_reference_document,file_attachment.description,file_attachment.is_custom,file_attachment.is_group,new_parent_attachment,file,file_attachment.is_required)
    return new_attachment






